### ISSUE TITLE
My sourcegraph docker install has difficulty cloning my GitLab repos

### ISSUE THREAD

**Issue by Tercan-Edmunds** [sourcegraph/sourcegraph#699]
- **Sourcegraph version: 2.12.3** 
- **Platform information: Linux (4.18.16-arch1), Docker: 18.06.1-ce** 

#### Steps to reproduce:

1. Setup GitLab Repository in Site Config, with correct URL and API Token
2. Go to Respository List and enable one of the repos

#### Expected behavior:
- The repo is cloned and indexed

#### Actual behavior:
- The repo reports "Not Cloned" for a very long time (hours)

#### Log Example: (timestamps removed)
```
repo-updater | lvl=warn msg="Error updating repo metadata" repo=gitlab.vip.aws2/edmunds/consumers-inventory-elastic err="context canceled"
gitserver | lvl=warn msg="error cloning repo" repo=gitlab.vip.aws2/edmunds/consumers-inventory-elastic err="error cloning repo: repo gitlab.vip.aws2/edmunds/consumers-inventory-elastic (http://git:[TOKEN]@gitlab.vip.aws2/edmunds/consumers-inventory-elastic.git) not cloneable: signal: killed"
```

**Comment by nicksnyder** [comment]
@Tercan-Edmunds Sorry to hear that! Can you clarify if the repo ever clones? For example, is it slow or does it never work at all?

**Comment by Tercan-Edmunds** [comment]
A handful of repos will eventually clone (after a few hours), but the longer the queue, the longer the clone will take. In one install, with 100 repos, the last 10 had not been cloned after 72 hours...

**Comment by nicksnyder** [comment]
Thanks for the info.

The first log line comes from here:
https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/repo-updater/repoupdater/server.go#L116

This is an async request to update some metadata, and I suspect it is hitting some default timeout, so I don't think this log line is as interesting as the second one.

The second log line comes from here:
https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/gitserver/server/server.go#L763:59

We are trying to do a `git ls-remote` with a [5 second timeout](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/gitserver/server/server.go#L968) and that appears to be timing out (e.g. `signal: killed` is what happens when the context times out).

From your own dev machine (or ideally from the machine that Sourcegraph is running on) can you run the following command:

```
time git ls-remote http://git:[TOKEN]@gitlab.vip.aws2/edmunds/consumers-inventory-elastic.git HEAD
```

If that takes close to or longer than 5 seconds we may just be hitting an aggressive timeout.

**Comment by Tercan-Edmunds** [comment]
Results in
```
git ls-remote 0.02s user 17.91s system 83% cpu 21.371 total.
```
(There are A LOT of tags due to our continuous integration system)

**Comment by Tercan-Edmunds** [comment]
Can the timeout be configurable?

**Comment by nicksnyder** [comment]
It is not currently possible to configure that timeout. Let me think about if there is a better command that we could run which would always be fast.

**Comment by nicksnyder** [comment]
@Tercan-Edmunds Does the command speed up if you provide the `--heads` tag?

```
time git ls-remote --heads http://git:[TOKEN]@gitlab.vip.aws2/edmunds/consumers-inventory-elastic.git HEAD
```

**Comment by Tercan-Edmunds** [comment]
actually, missed the HEAD in your original ls-remote command...
The total time bounces around 2.5s to 6 on the host.

In container --heads is significantly slower between 10-15s
In container no --heads runs between 5.5-9s

**Comment by nicksnyder** [comment]
Ah, looks like `ls-remote` does filtering locally...but there is [a recent improvement in git 2.18](https://www.infoq.com/news/2018/06/git-2.18-v2-protocol-commitgraph) that might make this better:

The version of git in the Sourcegraph container should already be >=2.18 so can you try this:

```
# without --heads
time git -c protocol.version=2 ls-remote http://git:[TOKEN]@gitlab.vip.aws2/edmunds/consumers-inventory-elastic.git HEAD

# with --heads
time git -c protocol.version=2 ls-remote --heads http://git:[TOKEN]@gitlab.vip.aws2/edmunds/consumers-inventory-elastic.git HEAD
```

**Comment by Tercan-Edmunds** [comment]
In Container:
Without Heads [2.66,8.33,4.13,2.00,2.31] Avg: 3.87s
With Heads [3.64,3.22,3.75,9.02,11.43] Avg: 6.21s (and doesn't return anything like w/o --heads)

**Comment by nicksnyder** [comment]
I just looked up GitLab docs and it looks like you would need GitLab 11.4 ([released October 22](https://about.gitlab.com/2018/10/22/gitlab-11-4-released/)) to take advantage of the wire protocol version 2:
https://docs.gitlab.com/ee/administration/git_protocol.html

I assume you haven't upgraded to 11.4 yet? If not, you should push to upgrade since it would also benefit your devs when they run things like `git fetch` since your repos have so many refs.

In the meantime I have opened a PR to propose increasing the timeout for `ls-remote`.
https://github.com/sourcegraph/sourcegraph/pull/757

**Comment by nicksnyder** [comment]
This fix is released in 2.13.1
